source/extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -150,6 +161,37 @@ Network::IoResult SslSocket::doRead(Buffer::Instance& read_buffer) {
   return {action, bytes_read, end_stream};
 }
 
+void SslSocket::onPrivateKeyMethodComplete() {
+  ASSERT(isThreadSafe());
+  ASSERT(async_handshake_in_progress_);
+  async_handshake_in_progress_ = false;
+
+  ENVOY_CONN_LOG(debug, "async handshake complete", callbacks_->connection());
+  if (!handshake_complete_) {
+    // It's possible that the async call comes in later, but the handshake has been retried from
+    // doWrite or similar. In that case the error handling has been done synchronously. */
+    PostIoAction result = doHandshake();
+    if (async_handshake_in_progress_ || result != PostIoAction::KeepOpen) {
+      // It may be semantically possible that the handshake is expecting read() or write(), so
+      // we shouldn't require that the handshake is complete. However, there can't be a second
+      // private key operation requested.
+      if (callbacks_->connection().state() == Network::Connection::State::Open) {
+        // The connection state machine thinks we are still connecting, but the second part of the
+        // private key method handshake failed. If the connection state would be something else,
+        // the socket was already closed while waiting for async handshake.
+        ENVOY_CONN_LOG(debug, "async handshake completion error", callbacks_->connection());
+        drainErrorQueue();
```

Comment:
`drainErrorQueue()` is already called in the only codepath that returns `PostIoAction::Close`.